### PR TITLE
Web3 sha3 handling input

### PR DESCRIPTION
**Pull request by heitorfm**
## Description
Improve compatibility with accepting only hexa encoded inputs for the sha3 hash
https://github.com/rsksmart/rskj/issues/1607

## Motivation and Context
compatibility with web3.js

## How Has This Been Tested?
unit tests

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+import static org.ethereum.rpc.exception.RskJsonRpcRequestException.invalidParamError;
+
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.util.regex.Pattern;
+
+/**
+ * Hex utils
+ */
+public class HexUtils {
```
[the remark this answers is not in the document]
Comment (heitorfm) [reply rskj-core/src/main/java/co/rsk/util/HexUtils.java:34]: This Class is changing location. In the near future, `TypeConverter `will be deleted.

File: rskj-core/src/main/java/org/ethereum/rpc/TypeConverter.java [hunk 2]
```diff
@@ -28,7 +28,9 @@
 
 /**
  * Created by Ruben on 19/11/2015.
+ * @deprecated in favor of {@link co.rsk.util.HexUtils HexUtils}.
  */
+@Deprecated
```
[the remark this answers is not in the document]
Comment (heitorfm) [reply rskj-core/src/main/java/org/ethereum/rpc/TypeConverter.java:33]: To change the package and naming for something more meaningful.

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    	
+    	return true;
+    }
+    
+    /**
+     * if string is hexadecimal with 0x prefix
+     */
+	public static boolean isHexWithPrefix(final String data) {
+		
+		if(data == null) {
```
[the remark this answers is not in the document]
Comment (heitorfm) [reply rskj-core/src/main/java/co/rsk/util/HexUtils.java:182]: done

File: rskj-core/src/main/java/org/ethereum/rpc/Web3Impl.java [hunk 4]
```diff
@@ -201,15 +195,29 @@ public String web3_clientVersion() {
...
+    	
+        String hash = null;
+        
         try {
-            byte[] result = HashUtil.keccak256(data.getBytes(StandardCharsets.UTF_8));
-            return s = TypeConverter.toJsonHex(result);
+        	
+        	if(HexUtils.isHexWithPrefix(data)) {
+        		
+        		byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
```
[the remark this answers is not in the document]
Comment (heitorfm) [reply rskj-core/src/main/java/org/ethereum/rpc/Web3Impl.java:205]: done

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+    		}
+    	}
+    	
+    	return true;
+    }
+    
+    /**
+     * if string is hexadecimal with 0x prefix
+     */
+	public static boolean isHexWithPrefix(final String data) {
```
[the remark this answers is not in the document]
Comment (heitorfm) [reply rskj-core/src/main/java/co/rsk/util/HexUtils.java:180]: `Utils.isHexadecimalString()` uses regexp which is the slowest possible way to do it. In the next steps, we need to clear out those hexa handling from `Util` to `HexUtil` as well.

**Comment by heitorfm** [comment]
> Hi I'll give my two cents. I've tried to do a basic curl and it gives me a wrong output
> 
> ```
> ~/rskj web3-sha3-handling-input ❯ curl http://localhost:4444 \
>   -s \
>   -X POST \
>   -H "Content-Type: application/json" \
>   --data '{"jsonrpc":"2.0","method":"web3_sha3","params":["62617A2875696E7433322C626F6F6C29"],"id":1}'
> {"jsonrpc":"2.0","id":1,"error":{"code":-32602,"message":"Parameter must be hexadecimal encoded with the '0x' prefix."}}
> 
> ~/rskj web3-sha3-handling-input ❯ curl http://localhost:4444 \
>   -s \
>   -X POST \
>   -H "Content-Type: application/json" \
>   --data '{"jsonrpc":"2.0","method":"web3_sha3","params":["0x62617A2875696E7433322C626F6F6C29"],"id":1}'
> {"jsonrpc":"2.0","id":1,"result":"0xcdcd77c0992ec5bbfc459984220f8c45084cc24d9b6efed1fae540db8de801d2"}
> ```
> 
> I expected it to return:
> 
> `{ "id":1, "jsonrpc": "2.0", "result": "0x47173285a8d7341e5e972fc677286384f802f8ef42a5ec5f03bbfa254cb01fad"}`
> 
> > Note: web3_sha3 takes in a hexidecimal number, not a direct string. So, if you wanted to convert "hello world" to it's Keccak-256 hash you would need to input the hex number for "hello world", which is "68656c6c6f20776f726c64".
> 
> * [Link 1](http://man.hubwiz.com/docset/Ethereum.docset/Contents/Resources/Documents/web3_sha3.html)
> * [Link 2](https://docs.alchemy.com/alchemy/apis/ethereum/web3_sha3)

Looks like you had the inputs and outputs mixed from the two links. You used the input of the second expecting the result of the first. And they are diferent.

File: rskj-core/src/main/java/org/ethereum/rpc/TypeConverter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Vovchyk) [reply rskj-core/src/main/java/org/ethereum/rpc/TypeConverter.java:33]: why to change the package?

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+
+    private static final byte[] HEX_PREFIX_BYTE_ARRAY = HEX_PREFIX.getBytes();
+
+    private static final byte[] ZERO_BYTE_ARRAY = "0".getBytes();
+    
+    private HexUtils() {
+        throw new IllegalAccessError("Utility class");
+    }
+
+    public static BigInteger stringNumberAsBigInt(String input) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:48]: please add `javadoc`'s for public methods

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+        return new BigInteger(hexa, 16);
+    }
+
+    public static byte[] stringHexToByteArray(final String param) {
+        
+        if (param == null) {
+            throw invalidParamError("input cant be null");
+        }
+    	
+    	String result = removeHexPrefix(param);
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:70]: please align indentment in this file

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+        }
+
+        return result;
+    }
+
+    public static String toJsonHex(Coin x) {
+        return x != null ? x.asBigInteger().toString() : "" ;
+    }
+
+    public static String toJsonHex(String x) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:96]: could you add a test for this?

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+        return HEX_PREFIX + x;
+    }
+
+    /**
+     * @return A Hex representation of n WITHOUT leading zeroes
+     */
+    public static String toQuantityJsonHex(long n) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:103]: could you add a test for this?

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+        return HEX_PREFIX + Long.toHexString(n);
+    }
+
+    /**
+     * @return A Hex representation of n WITHOUT leading zeroes
+     */
+    public static String toQuantityJsonHex(BigInteger n) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:110]: and for this one?

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * if string is hexadecimal with 0x prefix
+     */
+    public static boolean isHexWithPrefix(final String data) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:180]: could you cover this with a tests?

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+        }
+
+        // start testing after 0x prefix
+        return isHex(value, 2);
+    }
+
+	/**
+	 * if a string is composed solely of hexa chars
+	 */
+    public static boolean isHex(final String data) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:203]: missing test

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+        return isHex(data, 0);
+    }
+
+	/**
+	 * if a string is composed solely of hexa chars
+	 * Starting at the given index
+	 */
+    public static boolean isHex(final String data, int startAt) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:211]: and here

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+        System.arraycopy(HEX_PREFIX_BYTE_ARRAY, 0, result, 0, HEX_PREFIX_BYTE_ARRAY.length);
+        System.arraycopy(encoded, 0, result, HEX_PREFIX_BYTE_ARRAY.length, encoded.length);
+
+        return result;
+    }
+
+    /**
+     * remove Hex Prefix from string
+     */
+	public static String removeHexPrefix(final String data) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:250]: missing tests

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+        if (result != null && result.length % 2 != 0) {
+            result = new byte[data.length + ZERO_BYTE_ARRAY.length];
+            System.arraycopy(ZERO_BYTE_ARRAY, 0, result, 0, ZERO_BYTE_ARRAY.length);
+            System.arraycopy(data, 0, result, ZERO_BYTE_ARRAY.length, data.length);
+        }
+
+        return result;
+	}
+
+    public static byte[] decode(byte[] dataBytes) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:282]: missing tests

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+        return Hex.decode(HexUtils.leftPad(HexUtils.removeHexPrefix(dataBytes)));
+    }
+
+    public static int jsonHexToInt(final String param) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:286]: please cover this with tests

**Review by Vovchyk — CHANGES_REQUESTED** [review]

File: rskj-core/src/main/java/org/ethereum/rpc/TypeConverter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (heitorfm) [reply rskj-core/src/main/java/org/ethereum/rpc/TypeConverter.java:33]: Better organization, its just a small refactory.

[the remark this answers is not in the document]
Comment (Vovchyk) [reply rskj-core/src/main/java/org/ethereum/rpc/TypeConverter.java:33]: I see. My concern is that the class is being used in multiple places, so this should be done with proper precaution, please cover it with tests and javadocs

[the remark this answers is not in the document]
Comment (heitorfm) [reply rskj-core/src/main/java/org/ethereum/rpc/TypeConverter.java:33]: done

**Comment by heitorfm** [comment]
pipeline:run

**Comment by nagarev** [comment]
pipeline:run

**Comment by nagarev** [comment]
pipeline:run

**Comment by nagarev** [comment]
pipeline:run

**Comment by nagarev** [comment]
pipeline:run

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+     */
+    public static boolean isHexWithPrefix(final String data) {
+
+        if (data == null) {
+            return false;
+        }
+
+        String value = data.toLowerCase();
+
+        if (value.startsWith("-")) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:219]: why do we need this check?

**Review by Vovchyk — APPROVED** [review]

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java (hunk 17 above)
Comment (heitorfm) [reply rskj-core/src/main/java/co/rsk/util/HexUtils.java:219]: hexa may come with negative sign

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+        if (result.length() % 2 != 0) {
+            result = ZERO_STR + result;
+        }
+        return Hex.decode(result);
+    }
+
+    /**
+     * String to byte array 
+     */
+    public static byte[] stringToByteArray(String input) {
```
Comment (joaquinlpereyra) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:97]: How is this related to hex numbers?

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+    }
+
+    /**
+     * Converts a byte array to a string according to ethereum json-rpc specifications
+     */
+    public static String toJsonHex(final byte[] param) {
+        String result = toUnformattedJsonHex(param);
+
+        if (HEX_PREFIX.equals(result)) {
+            return "0x00";
```
Comment (joaquinlpereyra) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:108]: This is not according to spec, should `return 0x` for unformatted data: https://eth.wiki/json-rpc/API

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+        }
+
+        return result;
+    }
+
+    /**
+     * Convert from Coin to its BigInteger value in String representation or empty if the parameter is null
+     */
+    public static String toJsonHex(final Coin coin) {
+        return coin != null ? coin.asBigInteger().toString() : "" ;
```
Comment (joaquinlpereyra) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:118]: I don't think this is up to spec either, should be "0x0", right?

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java (hunk 17 above)
Comment (joaquinlpereyra) [reply rskj-core/src/main/java/co/rsk/util/HexUtils.java:219]: Negative hex should not be valid in JSON-RPC as the two formats are quantities or raw bytes, right?

**Review by joaquinlpereyra — COMMENTED** [review]
I left a couple of comments where I believe there are errors according to the spec.

But on a more general note, shouldn't this code be simpler? As far as as I see it, a `JSONRPCHex` class should have only four (public) functions. 

- `encodeQuantity(bigNumber)` and its inverse
- `encodeRaw(bytes)` and its inverse 

That would make it a 1:1 with the specification and much more easier to understand / audit. Note that in practice these are the functions that are used mostly used in practice, where the other's are probably found in other packages.

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+
+    /**
+     * if the parameter has the hex prefix 
+     */
+    public static boolean hasHexPrefix(final byte[] data) {
+
+        if (data == null) {
+            return false;
+        }
+
```
Comment (VaWheel) [inline rskj-core/src/main/java/co/rsk/util/HexUtils.java:198]: should we check `data` length here instead of checking it in the caller?

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java (hunk 17 above)
Comment (heitorfm) [reply rskj-core/src/main/java/co/rsk/util/HexUtils.java:219]: deleted the treatment

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java (hunk 19 above)
Comment (heitorfm) [reply rskj-core/src/main/java/co/rsk/util/HexUtils.java:108]: done

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java (hunk 21 above)
Comment (heitorfm) [reply rskj-core/src/main/java/co/rsk/util/HexUtils.java:198]: yes

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java (hunk 19 above)
Comment (heitorfm) [reply rskj-core/src/main/java/co/rsk/util/HexUtils.java:108]: It was being done this way before. Are we gona change it now ?

**Comment by heitorfm** [comment]
> I left a couple of comments where I believe there are errors according to the spec.
> 
> But on a more general note, shouldn't this code be simpler? As far as as I see it, a `JSONRPCHex` class should have only four (public) functions.
> 
>     * `encodeQuantity(bigNumber)` and its inverse
> 
>     * `encodeRaw(bytes)` and its inverse
> 
> 
> That would make it a 1:1 with the specification and much more easier to understand / audit. Note that in practice these are the functions that are used mostly used in practice, where the other's are probably found in other packages.

This class, besides small additions to the web3_sha3 task is a refactoring from `TypeConverter` which iḿ moving from package and changing name for a more meaningful one.

File: rskj-core/src/main/java/co/rsk/util/HexUtils.java (hunk 20 above)
Comment (heitorfm) [reply rskj-core/src/main/java/co/rsk/util/HexUtils.java:118]: This, was already present in TypeConverter

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Review by Vovchyk — APPROVED** [review]
